overlord/snapstate/snapstate.go

```diff
@@ -273,6 +274,89 @@ func isCoreSnap(snapName string) bool {
 	return snapName == defaultCoreSnapName
 }
 
+// removeExtraComponentsTasks creates tasks that will remove unwanted components
+// that are currently installed alongside the snap revision that is about to be
+// installed. If the new snap revision is not in the sequence, then we don't
+// have anything to do. If the revision is in the sequence, then we generate
+// tasks that will unlink components that are not in compsups.
+//
+// This is mostly relevant when we're moving from one snap revision to another
+// snap revision that has already been installed on the system. The target snap
+// might have components that are installed that we don't want any more.
+func removeExtraComponentsTasks(st *state.State, snapst *SnapState, newRevision snap.Revision, compsups []ComponentSetup) (
```

Comment:
newRevision is a bit confusing, as it might be an old revision, at least in terms of being on the system which is actually the case where this does something